File: src/main/java/designPatterns/builder/BuilderRunner.java
package designPatterns.builder;

import designPatterns.builder.notRequiredProperties.NoPropsRequired;
import designPatterns.builder.reflection.BuilderForSampleObj;
import designPatterns.builder.reflection.SampleObj;
import designPatterns.builder.reflectionBasedProxy.ReflectionBuilderFactory;
import designPatterns.builder.requiredProperties.AllPropsRequired;

import java.lang.reflect.Method;

public class BuilderRunner {
    public static void main(String[] args) {
        AllPropsRequired allPropsRequired = AllPropsRequired.builder()
                .age(15)
                .name("name")
                .build();

        NoPropsRequired noPropsRequired = NoPropsRequired.builder().build();

        ReflectionBuilderFactory reflectionBuilderFactory = new ReflectionBuilderFactory();


        Method[] methods = BuilderForSampleObj.class.getMethods();

        BuilderForSampleObj instance = reflectionBuilderFactory.getInstance(BuilderForSampleObj.class, SampleObj.class);

        SampleObj sampleObj = instance
                .name("name")
                .value(100)
                .build();

        System.out.println(sampleObj.getName());
        System.out.println(sampleObj.getValue());
        System.out.println(sampleObj);

    }

}

File: src/main/kotlin/designPatterns/builder/BuilderRunner.kt
package designPatterns.builder

import designPatterns.builder.notRequiredProperties.KotlinBuilderNonRequired
import designPatterns.builder.requiredProperties.KotlinBuilderAllRequired

fun main(args: Array<String>) {
    val allRequired = KotlinBuilderAllRequired("name", 100)

    println("allRequired -> $allRequired")

    val allDefault = KotlinBuilderNonRequired()
    val nameSet = KotlinBuilderNonRequired("not the default name")
    val ageSet = KotlinBuilderNonRequired(age = 100)
    val bothSetOutOfOrder = KotlinBuilderNonRequired(
            age = 100,
            name = "new name"
    )

    val kotlinBuilderNonRequired = KotlinBuilderNonRequired(age = 12, name = "some string", another = "string")
    println(kotlinBuilderNonRequired.age)

    println("allDefault -> $allDefault")
    println("nameSet -> $nameSet")
    println("ageSet -> $ageSet")
    println("bothSetOutOfOrder -> $bothSetOutOfOrder")

    val kotlinBuilderAllRequired = KotlinBuilderAllRequired(
            name = "some name"
    )
}